app/src/main/java/com/github/geohunt/app/data/repository/ClaimRepository.kt

```diff
@@ -0,0 +1,184 @@
+                }.awaitAll().zip(this).map {
+                    if (!it.first.exists()) {
+                        throw ClaimNotFoundException(it.second)
+                    }
+                    it.first.getValue(FirebaseClaim::class.java)!!.asExternalModel(it.second)
+                }
+            }
+        }
+    }
+
+    /**
+     * Retrieve a list of all claims associated with the current challenges
+     */
+    override suspend fun getClaimsByChallenge(challenge: Challenge): List<Claim> {
+        require(challenge.id.isNotEmpty())
+
+        return withContext(ioDispatcher) {
+            database.getReference("claims/${challenge.id}")
+                .get()
+                .await()
+                .run {
+                    children.mapNotNull {
+                        val value = it.getValue(FirebaseClaim::class.java)
+                        if (value != null) it.key!! to value
+                        else null
+                    }
+                    .map {
+                        it.second.asExternalModel("${challenge.id}/${it.first}")
+                    }
+                }
+        }
+    }
+
+    /**
+     * Claim a specific challenge with the given photo and location
+     *
+     * The claim will be claimed on the behalf of the currently logged used. If there
+     * is none, throws a [UserNotLoggedInException]
+     */
+    @Throws(UserNotLoggedInException::class)
+    override suspend fun claimChallenge(
+        photo: LocalPicture,
+        location: Location,
+        challenge: Challenge
+    ): Claim = withContext(ioDispatcher) {
+        authRepository.requireLoggedIn()
+
+        val currentUser = authRepository.getCurrentUser()
+
+        val claimRef = database.getReference("claims/${challenge.id}").push()
+        val claimByUser = database.getReference("claimsByUser/${currentUser.id}").push() // Notice that the key here make no sense
+        val claimId = challenge.id + "/" + claimRef.key!!
+
+        // First upload the image to Firebase storage
+        // This ensures that the database doesn't contain nonexistent image data
+        val photoUrl: Uri = imageRepository.uploadClaimPhoto(photo, claimId)
+
+        // Compute the distance to the target
+        val distance = location.distanceTo(challenge.location)
+
+        // Upload the entry to Firebase's Realtime Database
+        val claimEntry = FirebaseClaim(
+            currentUser.id,
+            time = DateUtils.utcIso8601Now(),
+            photoUrl = photoUrl.toString(),
+            cid = challenge.id,
+            location = location,
+            distance = (distance.toLong() + 1),
+            awardedPoints = pointCalculatorMap[challenge.difficulty]!!.computePoints(distance)
+        )
+
+        // Upload the entry
+        awaitAll(
+            claimRef.setValue(claimEntry).asDeferred(),
+            claimByUser.setValue(claimId).asDeferred()
+        )

```

Comment:
Can be replaced with simple `await` calls:
```suggestion
        claimRef.setValue(claimEntry).await()
        claimByUser.setValue(claimId).await()
```

Reply:
Same issue as above this is sequential and therefore slower